File: 01/main.hs
module Main (main) where

import System.IO

main :: IO ()
main = do
    depthData <- readFile "./01/depths.txt"
    let depths = map (\x -> read x :: Int) (lines depthData)
    putStrLn $ "Depth increases: " ++ show (increases depths)
    putStrLn $ "3-Windowed increases: " ++ show (windowedIncreases 3 depths)

bump :: Bool -> Int
bump true = if true then 1 else 0

increases :: [Int] -> Int
increases (x:xs) =
    fst $ foldl (\(inc, a) b -> (inc + (bump $ b > a), b)) (0, x) xs

windowed :: Int -> [a] -> [[a]]
windowed _ [] = []
windowed size ls =
    if length ls >= size
        then (take size ls) : windowed size (tail ls)
        else windowed size (tail ls)

windowedIncreases :: Int -> [Int] -> Int
windowedIncreases size xs =
  let
    (w:ws) = map (foldr (+) 0) (windowed size xs)
  in
    fst $ foldl (\(inc, a) b -> (inc + (bump $ b > a), b)) (0, w) ws

File: 01/main.ts
const decoder = new TextDecoder("utf-8");
const depthData = await Deno.readFile("./01/depths.txt");

const depths = decoder.decode(depthData).split("\n").map((s) => parseInt(s));

// Part One
const [increases] = depths.slice(1).reduce(
  ([inc, y], x) => [inc + (x > y ? 1 : 0), x],
  [0, depths[0]],
);
console.log(`Increases: ${increases}`);

// Part Two
const [windowedIncreases] = depths.slice(1, -3).reduce(
  ([inc, y], x, i) => {
    const sum = x + depths[i + 2] + depths[i + 3];
    return [inc + (sum > y ? 1 : 0), sum];
  },
  [0, depths.slice(0, 3).reduce((y, x) => x + y)],
);
console.log(`3-Windowed Increases: ${windowedIncreases}`);
